<HTML>
<HEAD>
<TITLE>Exceptions</TITLE>
</HEAD>
<BODY bgcolor="#FFFFFF">
 <img src="..\icons\cover75.gif" alt="Cover" align=right>
<b>Data Structures and Algorithms 
with Object-Oriented Design Patterns in Java</b><br>
<A NAME="tex2html9464" HREF="page615.html"><IMG WIDTH=37 HEIGHT=24 ALIGN=BOTTOM ALT="next" SRC="..\icons\next_motif.gif"></A> <A NAME="tex2html9462" HREF="page585.html"><IMG WIDTH=26 HEIGHT=24 ALIGN=BOTTOM ALT="up" SRC="..\icons\up_motif.gif"></A> <A NAME="tex2html9458" HREF="page613.html"><IMG WIDTH=63 HEIGHT=24 ALIGN=BOTTOM ALT="previous" SRC="..\icons\previous_motif.gif"></A> <A NAME="tex2html9466" HREF="page9.html"><IMG WIDTH=65 HEIGHT=24 ALIGN=BOTTOM ALT="contents" SRC="..\icons\contents_motif.gif"></A> <A NAME="tex2html9467" HREF="page618.html"><IMG WIDTH=43 HEIGHT=24 ALIGN=BOTTOM ALT="index" SRC="..\icons\index_motif.gif"></A> <BR><HR>
<H1><A NAME="SECTION0018600000000000000000">Exceptions</A></H1>
<A NAME="secjavaexceptions">&#160;</A>
<P>
Sometimes unexpected situations arise during the execution of a program.
Careful programmers write code that detects errors and
deals with them appropriately.
However, a simple algorithm can become unintelligible
when error-checking is added because the error-checking code
can obscure the normal operation of the algorithm.
<P>
<em>Exceptions</em><A NAME=56934>&#160;</A> provide a clean way to detect
and handle unexpected situations.
When a program detects an error,
it <em>throws</em><A NAME=56936>&#160;</A> an exception.
When an exception is thrown,
control is transfered to the appropriate
<em>exception handler</em><A NAME=56938>&#160;</A>.
By defining a method that <em>catches</em> the exception,
the programmer can write the code to handle the error.
<P>
In Java, an exception is an object.
All exceptions in Java are ultimately derived from the base
class called <tt>java.lang.Throwable</tt>.
For example,
consider the class <tt>A</tt> defined in Program&nbsp;<A HREF="page614.html#progExamplex"><IMG  ALIGN=BOTTOM ALT="gif" SRC="..\icons\cross_ref_motif.gif"></A>.
Since the <tt>A</tt> class extends the <tt>Throwable</tt> class,
<tt>A</tt> is an exception that can be <em>thrown</em>.
<P>
<P><A NAME="56950">&#160;</A><A NAME="progExamplex">&#160;</A> <IMG WIDTH=575 HEIGHT=390 ALIGN=BOTTOM ALT="program56947" SRC="img2514.gif"  ><BR>
<STRONG>Program:</STRONG> Using exceptions in Java.<BR>
<P>
<P>
A method throws an exception by using the <tt>throw</tt> statement:
The <tt>throw</tt> statement is similar to a <tt>return</tt> statement.
A <tt>return</tt> statement represents the normal termination
of a method and the object returned matches the return value of the method.
A <tt>throw</tt> statement represents the abnormal termination of a method
and the object thrown represents the type of error encountered.
<P>
The <tt>f</tt> method in Program&nbsp;<A HREF="page614.html#progExamplex"><IMG  ALIGN=BOTTOM ALT="gif" SRC="..\icons\cross_ref_motif.gif"></A> throws an <tt>A</tt> exception.
Java requires that a method which throws exceptions
that are not caught within that method (see below)
must declare all the exceptions it throws
unless those exceptions are derived from the special classes
<tt>java.lang.Error</tt><A NAME=57041>&#160;</A>
or <tt>java.lang.RuntimeException</tt><A NAME=57042>&#160;</A>.<A NAME="tex2html1208" HREF="footnode.html#57043"><IMG  ALIGN=BOTTOM ALT="gif" SRC="..\icons\foot_motif.gif"></A>
Thus, the <tt>f</tt> method declares that it throws <tt>A</tt>.
<P>
Exception handlers are defined using a <tt>try</tt> block:
The body of the <tt>try</tt> block is executed
either until an exception is thrown or until it terminates normally.
One or more exception handlers follow a <tt>try</tt> block.
Each exception handler consists of a <tt>catch</tt> clause
which specifies the exceptions to be caught,
and a block of code,
which is executed when the exception occurs.
When the body of the <tt>try</tt> block throws an exception
for which an exception is defined,
control is transfered to the body of the exception handler.
<P>
In this example, the exception thrown by the <tt>f</tt> method
is caught by the <tt>g</tt> method.
In general when an exception is thrown,
the chain of methods called is searched in reverse
(from caller to callee)
to find the closest matching <tt>catch</tt> statement.
When a program throws an exception that is not caught,
the program terminates.
<P>
<HR><A NAME="tex2html9464" HREF="page615.html"><IMG WIDTH=37 HEIGHT=24 ALIGN=BOTTOM ALT="next" SRC="..\icons\next_motif.gif"></A> <A NAME="tex2html9462" HREF="page585.html"><IMG WIDTH=26 HEIGHT=24 ALIGN=BOTTOM ALT="up" SRC="..\icons\up_motif.gif"></A> <A NAME="tex2html9458" HREF="page613.html"><IMG WIDTH=63 HEIGHT=24 ALIGN=BOTTOM ALT="previous" SRC="..\icons\previous_motif.gif"></A> <A NAME="tex2html9466" HREF="page9.html"><IMG WIDTH=65 HEIGHT=24 ALIGN=BOTTOM ALT="contents" SRC="..\icons\contents_motif.gif"></A> <A NAME="tex2html9467" HREF="page618.html"><IMG WIDTH=43 HEIGHT=24 ALIGN=BOTTOM ALT="index" SRC="..\icons\index_motif.gif"></A> <P><ADDRESS>
<img src="..\icons\bruno.gif" alt="Bruno" align=right>
<a href="..\copyright.html">Copyright &#169; 1998</a> by <a href="..\signature.html">Bruno R. Preiss, P.Eng.</a>  All rights reserved.

</ADDRESS>
</BODY>
</HTML>
